### ISSUE TITLE
"You must pass the size of each sequence in the batch as a tensor" bug when running.

### ISSUE THREAD

**Issue by NgPDat** [SeanNaren/CTCNetworkClient#23]
Here is the full error message:

/opt/zbstudio/bin/linux/x64/lua: /home/pdat/torch/install/share/lua/5.1/nnx/CTCCriterion.lua:24: You must pass the size of each sequence in the batch as a tensor
stack traceback:
    [C]: in function 'assert'
    /home/pdat/torch/install/share/lua/5.1/nnx/CTCCriterion.lua:24: in function 'forward'
    ./Network.lua:79: in function 'opfunc'
    /home/pdat/torch/install/share/lua/5.1/optim/sgd.lua:44: in function 'sgd'
    ./Network.lua:102: in function 'trainNetwork'
    ...dat/torch/workspace/CTCSpeechRecognition/AN4CTCTrain.lua:47: in main chunk
    [C]: at 0x004054e8

Then I check CTCCriterion.lua in nnx package, see it requires 3 parameters to the 'forward' function.
Any help to fix this issue is very appreciated.

**Comment by SeanNaren** [comment]
Hey,

A temporary fix for this would be to change the local function 'feval' to look like this in the Network.lua class:

From this:

```
    local function feval(x_new)
        local inputsCPU, targets = dataset:nextData()
        -- transfer over to GPU
        inputs:resize(inputsCPU:size()):copy(inputsCPU)
        gradParameters:zero()
        local predictions = self.model:forward(inputs)
        local loss = ctcCriterion:forward(predictions, targets)
        self.model:zeroGradParameters()
        local gradOutput = ctcCriterion:backward(predictions, targets)
        self.model:backward(inputs, gradOutput)
        return loss, gradParameters
    end
```

To this:

```
    local function feval(x_new)
        local inputsCPU, targets = dataset:nextData()
        -- transfer over to GPU
        inputs:resize(inputsCPU:size()):copy(inputsCPU)
        gradParameters:zero()
        local predictions = self.model:forward(inputs)
        local tensorSizes = predictions:size()
        local sizes = torch.Tensor(tensorSizes[1]):fill(tensorSizes[2])
        local loss = ctcCriterion:forward(predictions, targets, sizes)
        self.model:zeroGradParameters()
        local gradOutput = ctcCriterion:backward(predictions, targets, sizes)
        self.model:backward(inputs, gradOutput)
        return loss, gradParameters
    end
```

I'm going to change master to have this in temporarily now, thanks!

**Comment by NgPDat** [comment]
Your fix solved the problem. Thank you!

**Comment by clarkong** [comment]
hi @SeanNaren , I met this problem too, and fixed it as you said. But my training seems to be not converged, the `Average Loss` is 900+ and `WER` is 100% after 70 epochs.

![20160521120442](https://cloud.githubusercontent.com/assets/8656702/15446250/3f81e1fa-1f4c-11e6-8a8b-9dfa2ec5162b.png)

do you have any idea?

**Comment by SeanNaren** [comment]
@clarkong sorry for taking long to respond, have you done any modifications to the base code at all?

**Comment by clarkong** [comment]
I just modify function `feval` as you said. and I have tried to  increase learning rate but the loss is `inf`. I doubt it's maybe some installation problems and I will try it on other machine.

**Comment by clarkong** [comment]
After pull master, all things go right :) thank you very much. @SeanNaren
